File: LeetCode/Easy/0563-binary-tree-tilt/0563-binary-tree-tilt.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTilt(self, root: Optional[TreeNode]) -> int:
        self.ret = 0
        def DFS(node):

            if not node:
                return 0

            left = DFS(node.left)
            right = DFS(node.right)

            self.ret += abs(left - right)

            return left + right + node.val

        DFS(root)
        
        return self.ret

File: LeetCode/Easy/0563-binary-tree-tilt/0563-binary-tree-tilt.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ret = 0;
    int findTilt(TreeNode* root) {
        DFS(root);
        return ret;
    }
    
    int DFS(TreeNode* node) {
        if (!node) return 0;
        
        int left = DFS(node->left);
        int right = DFS(node->right);
        ret += abs(left - right);
        
        return left + right + node->val;
    }
};